File: swift/Sources/Ensoul/Resources/Aggregate.swift
/// Aggregate resource for the Ensoul Swift SDK.
///
/// Wraps all `/v1/aggregate` endpoints, including the SSE streaming variants
/// used for real-time confidence-tracked aggregation.
///
/// Example — one-shot query:
/// ```swift
/// let result = try await client.aggregate.query("What do people think of remote work?")
/// ```
///
/// Example — streaming:
/// ```swift
/// let stream = client.aggregate.stream("Brand sentiment for product X")
/// for try await event in stream {
///     if let payload = try? JSONSerialization.jsonObject(with: event.data.data(using: .utf8)!) {
///         print(payload)
///     }
/// }
/// ```
import Foundation

// MARK: - Aggregate

@available(iOS 15.0, macOS 12.0, *)
public class Aggregate {
    private let client: HTTPClient

    public init(client: HTTPClient) {
        self.client = client
    }

    // MARK: - Count

    /// GET /v1/aggregate/count — count personas matching a filter.
    public func count(
        domain: String? = nil,
        filters: String? = nil,
        region: String? = nil,
        archetype: String? = nil,
        ageMin: Int? = nil,
        ageMax: Int? = nil
    ) async throws -> [String: Any] {
        var params: [String: String] = [:]
        if let domain { params["domain"] = domain }
        if let filters { params["filters"] = filters }
        if let region { params["region"] = region }
        if let archetype { params["archetype"] = archetype }
        if let ageMin { params["age_min"] = String(ageMin) }
        if let ageMax { params["age_max"] = String(ageMax) }

        let (data, _) = try await client.get("/v1/aggregate/count", params: params)
        return try Self.jsonObject(from: data)
    }

    // MARK: - Stats

    /// GET /v1/aggregate/stats — aggregate query statistics.
    public func stats() async throws -> [String: Any] {
        let (data, _) = try await client.get("/v1/aggregate/stats")
        return try Self.jsonObject(from: data)
    }

    // MARK: - Stream

    /// POST /v1/aggregate/stream
    ///
    /// Opens an SSE stream that emits progress events as the server processes
    /// personas. The stream terminates when `target_confidence` is reached or
    /// `max_samples` have been processed.
    ///
    /// This method is synchronous — no network connection is opened until the
    /// caller begins iterating the returned `SSEStream`.
    public func stream(
        _ query: String,
        filters: [String: Any]? = nil,
        aggregationMode: String? = nil,
        targetConfidence: Double = 0.95,
        minSamples: Int = 100,
        maxSamples: Int? = nil
    ) -> SSEStream {
        var body: [String: Any] = [
            "query": query,
            "target_confidence": targetConfidence,
            "min_samples": minSamples,
        ]
        if let filters { body["filters"] = filters }
        if let aggregationMode { body["aggregation_mode"] = aggregationMode }
        if let maxSamples { body["max_samples"] = maxSamples }

        return client.streamSSE(
            method: "POST",
            path: "/v1/aggregate/stream",
            body: body
        )
    }

    // MARK: - Grouped Stream

    /// POST /v1/aggregate/stream/grouped
    ///
    /// Like `stream`, but groups results by a specified field (e.g. `"region"`,
    /// `"age_range"`) and emits per-group progress events.
    ///
    /// This method is synchronous — no network connection is opened until the
    /// caller begins iterating the returned `SSEStream`.
    public func groupedStream(
        _ query: String,
        groupBy: String,
        filters: [String: Any]? = nil
    ) -> SSEStream {
        var body: [String: Any] = [
            "query": query,
            "group_by": groupBy,
        ]
        if let filters { body["filters"] = filters }

        return client.streamSSE(
            method: "POST",
            path: "/v1/aggregate/stream/grouped",
            body: body
        )
    }

    // MARK: - Simulate

    /// POST /v1/aggregate/simulation (`SimulationRequest`).
    ///
    /// Runs a forward simulation for a given scenario against an optional
    /// target cohort and returns the projected outcome as a raw JSON dictionary.
    public func simulate(
        scenario: String,
        targetCohort: [String: Any]? = nil,
        durationDays: Int = 30,
        parameters: [String: Any]? = nil
    ) async throws -> [String: Any] {
        var body: [String: Any] = [
            "scenario": scenario,
            "duration_days": durationDays,
        ]
        if let targetCohort { body["target_cohort"] = targetCohort }
        if let parameters { body["parameters"] = parameters }

        let (data, _) = try await client.post("/v1/aggregate/simulation", body: body)
        return try Self.jsonObject(from: data)
    }

    // MARK: - Trace Influence

    /// GET /v1/aggregate/influence/{personaId}
    ///
    /// Returns the influence graph for a persona, showing how its traits or
    /// interactions propagate through the population.
    public func traceInfluence(
        _ personaId: String,
        influenceType: String? = nil,
        direction: String = "downward",  // API accepts: downward | upward | both
        maxDepth: Int = 3
    ) async throws -> [String: Any] {
        var params: [String: String] = [
            "direction": direction,
            "max_depth": String(maxDepth),
        ]
        if let influenceType { params["influence_type"] = influenceType }

        let (data, _) = try await client.get(
            "/v1/aggregate/influence/\(personaId)",
            params: params
        )
        return try Self.jsonObject(from: data)
    }

    // MARK: - Private helpers

    private static func jsonObject(from data: Data) throws -> [String: Any] {
        guard let dict = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw EnsoulAPIError(
                statusCode: 200,
                error: "ParseError",
                message: "Expected a JSON object in aggregate response"
            )
        }
        return dict
    }
}

File: typescript/src/resources/aggregate.ts
import type { HTTPClient } from "../http.js";
import { SSEStream } from "../streaming.js";

export class Aggregate {
  constructor(private readonly client: HTTPClient) {}

  /** GET /v1/aggregate/count — count personas matching a filter. */
  async count(
    options: {
      domain?: string;
      filters?: string;
      region?: string;
      archetype?: string;
      ageMin?: number;
      ageMax?: number;
    } = {},
  ): Promise<Record<string, unknown>> {
    const params: Record<string, unknown> = {};
    if (options.domain != null) params.domain = options.domain;
    if (options.filters != null) params.filters = options.filters;
    if (options.region != null) params.region = options.region;
    if (options.archetype != null) params.archetype = options.archetype;
    if (options.ageMin != null) params.age_min = options.ageMin;
    if (options.ageMax != null) params.age_max = options.ageMax;
    return this.client.get("/v1/aggregate/count", params);
  }

  /** GET /v1/aggregate/stats — aggregate query statistics. */
  async stats(): Promise<Record<string, unknown>> {
    return this.client.get("/v1/aggregate/stats");
  }

  async stream(
    query: string,
    options: {
      filters?: Record<string, unknown>;
      aggregationMode?: string;
      targetConfidence?: number;
      minSamples?: number;
      maxSamples?: number;
    } = {},
  ): Promise<SSEStream> {
    const body: Record<string, unknown> = {
      query,
      target_confidence: options.targetConfidence ?? 0.95,
      min_samples: options.minSamples ?? 100,
    };
    if (options.filters != null) body.filters = options.filters;
    if (options.aggregationMode != null) body.aggregation_mode = options.aggregationMode;
    if (options.maxSamples != null) body.max_samples = options.maxSamples;
    const response = await this.client.streamSSE("/v1/aggregate/stream", body);
    return new SSEStream(response);
  }

  async groupedStream(
    query: string,
    options: { groupBy: string; filters?: Record<string, unknown> },
  ): Promise<SSEStream> {
    const body: Record<string, unknown> = { query, group_by: options.groupBy };
    if (options.filters != null) body.filters = options.filters;
    const response = await this.client.streamSSE("/v1/aggregate/stream/grouped", body);
    return new SSEStream(response);
  }

  async simulate(options: {
    scenario: string;
    targetCohort?: Record<string, unknown>;
    durationDays?: number;
    parameters?: Record<string, unknown>;
  }): Promise<Record<string, unknown>> {
    const body: Record<string, unknown> = {
      scenario: options.scenario,
      duration_days: options.durationDays ?? 30,
    };
    if (options.targetCohort != null) body.target_cohort = options.targetCohort;
    if (options.parameters != null) body.parameters = options.parameters;
    return this.client.post("/v1/aggregate/simulation", body);
  }

  async traceInfluence(
    personaId: string,
    options: {
      influenceType?: string;
      direction?: string;
      maxDepth?: number;
    } = {},
  ): Promise<Record<string, unknown>> {
    const params: Record<string, unknown> = {
      direction: options.direction ?? "downstream",
      max_depth: options.maxDepth ?? 3,
    };
    if (options.influenceType != null) params.influence_type = options.influenceType;
    return this.client.get(`/v1/aggregate/influence/${personaId}`, params);
  }
}
